### PR TITLE
Bootstrap Part 1: Small changes

### PR DESCRIPTION
**Pull request by chrisnojima**
This is a 3 part PR, only part 3 runs
- https://github.com/keybase/client/pull/13245
- https://github.com/keybase/client/pull/13246

The goal of these PRs is to cleanup and speed up the bootstrapping process.
A lot of the code this touches is very very old
This:

- [x] Simplifies push logic
- [x] Creates a more structured handshake/logout process so various sagas can register to do processing during that time correctly
- [x] Removes races in figuring out where to navigate to on start
- [x] Navigates directly to the target location
- [x] Removes the keys animation on resume
- [x] Uses quicker rpc to get provisioned usernames
- [x] Moves badge handling logic into various sagas instead of being localized in a single one
- [x] Parallelizes more things on startup
- [x] Removed some pointless extras paths in the login route structure
- [x] New colored logging for redux/engine

This PR focuses on smaller changes across a bunch of files

### CODE REVIEW

File: shared/actions/fs/platform-specific.desktop.js [hunk 1]
```diff
@@ -120,19 +119,10 @@ function* openWithCurrentMountDir(openPath: string): Saga.SagaGenerator<any, any
...
+  const kbfsPath = yield Saga.call(RPCTypes.kbfsMountGetCurrentMountDirRpcPromise)
 
   if (!kbfsPath) {
-    kbfsPath = yield Saga.call(RPCTypes.kbfsMountGetCurrentMountDirRpcPromise)
-
-    if (!kbfsPath) {
-      throw new Error('No kbfsPath (RPC)')
-    }
-
-    yield Saga.put(ConfigGen.createChangeKBFSPath({kbfsPath}))
```
Comment (chrisnojima) [inline shared/actions/fs/platform-specific.desktop.js]: not used

File: shared/actions/route-tree.js [hunk 2]
```diff
@@ -36,10 +36,10 @@ export function refreshRouteDef(
 
 // Switch the tree of route definitions. Dispatched when switching
 // from logged out to logged in and vice versa.
-export function switchRouteDef(routeDef: RouteDefParams): Types.SwitchRouteDef {
+export function switchRouteDef(routeDef: RouteDefParams, path?: Path): Types.SwitchRouteDef {
   return {
     type: Constants.switchRouteDef,
-    payload: {routeDef},
+    payload: {path, routeDef},
```
Comment (chrisnojima) [inline shared/actions/route-tree.js]: let us switch and set an initial path in one go

File: shared/app/main.android.js [hunk 3]
```diff
@@ -1,29 +0,0 @@
...
-import {compose, lifecycle, withProps} from 'recompose'
-import {NativeBackHandler} from '../common-adapters/mobile.native'
-import {getPath} from '../route-tree'
-
-module.hot &&
-  module.hot.accept(() => {
-    console.log('accepted update in main.android')
-  })
-
-export default compose(
```
Comment (chrisnojima) [inline shared/app/main.android.js]: this doesn't really do anything

File: shared/react-native/android/app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -27,7 +27,7 @@
       <activity
         android:name=".MainActivity"
         android:screenOrientation="portrait"
-        android:launchMode="singleTop"
+        android:launchMode="singleTask"
```
Comment (chrisnojima) [inline shared/react-native/android/app/src/main/AndroidManifest.xml]: i think this is more of what our style of app is. looking at the push issues this seems to make certain flows better. if we see weird android stuff we could revert this

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is big enough and important enough that this PR needs at least 2 thumbs . Please test part 3 on desktop and mobile. especially the push flows

File: shared/app/push-prompt.native.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      <Kb.Text type="Header" style={styles.text}>
+        Please turn on notifications!
+      </Kb.Text>
+      <Kb.RequireImage
+        style={styles.image}
+        src={require('../images/illustrations/illustration-turn-on-notifications-460-x-252.png')}
+      />
+      <Kb.Text type="BodySmallSemibold" style={styles.text}>
+        It's{' '}
+        <Kb.Text type="BodySmallSemiboldItalic" style={{color: Styles.globalColors.black}}>
```
Comment (buoyad) [inline shared/app/push-prompt.native.js]: Can factor out the style

File: shared/app/routes-login.js [hunk 6]
```diff
@@ -10,6 +10,7 @@ import loginRoutes from '../login/routes'
 const loginRouteTree = makeRouteDefNode({
   tags: makeLeafTags({title: loginRouteTreeTitle}),
   children: {
+    // $FlowIssue
```
Comment (buoyad) [inline shared/app/routes-login.js]: What's the problem?

File: shared/engine/index.js [hunk 7]
```diff
@@ -166,7 +166,12 @@ class Engine {
     const handlers = this._onDisconnectHandlers
     // Don't allow mutation while we're handling the handlers
```
Comment (buoyad) [inline shared/engine/index.js]: could this cause new handlers added while this is running to be dropped?

File: shared/engine/index.js [hunk 8]
```diff
@@ -427,8 +437,8 @@ class Engine {
     this._failOnError = true
   }
 
-  // Register a named callback when we disconnect from the server. Call if we're already disconnected
-  listenOnDisconnect(key: string, f: () => void) {
+  // Register a named callback when we disconnect from the server. Call if we're already disconnected. Callback should produce an action
```
Comment (buoyad) [inline shared/engine/index.js]: does `server` mean `service` here?

**Comment by buoyad** [comment]
higher level stuff:
* Inbox doesn't reload on logout -> login
* Linking to a profile in the app doesn't take you to the profile
* Push doesn't take me into the conversation on mobile

File: shared/actions/gregor.js [hunk 9]
```diff
@@ -35,57 +33,37 @@ function toNonNullGregorItems(state: GregorState): Array<Types.NonNullGregorItem
...
-        }
+const setupEngineListeners = () => {
+  // we get this with sessionID == 0 if we call openDialog
+  engine().setIncomingActionCreators('keybase.1.gregorUI.pushState', ({reason, state}, response) => {
+    response && response.result()
+    return [GregorGen.createPushState({reason, state})]
+  })
 
-        return actions
+  engine().setIncomingActionCreators('keybase.1.gregorUI.pushOutOfBandMessages', ({oobm}, response) => {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/gregor.js]: i'll ticket this, the generated code should do this

File: shared/actions/gregor.js [hunk 10]
```diff
@@ -96,23 +74,20 @@ function registerGregorListeners() {
...
-    const actions = []
-    if (oobm && oobm.length) {
-      const filteredOOBM = oobm.filter(oobm => !!oobm)
-      if (filteredOOBM.length) {
-        actions.push(GregorGen.createPushOOBM({messages: filteredOOBM}))
-      }
+  // The startReachibility RPC call both starts and returns the current
+  // reachability state. Then we'll get updates of changes from this state
+  // via reachabilityChanged.
+  // This should be run on app start and service re-connect in case the
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/gregor.js]: i'm hesitant to change this behavior in this pr

File: shared/app/main.native.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import {navigateUp, setRouteState} from '../actions/route-tree'
+import {GatewayDest} from 'react-gateway'
+import {NativeBackHandler} from '../common-adapters/mobile.native'
+import {View} from 'react-native'
+import {globalStyles} from '../styles'
+import {isAndroid} from '../constants/platform'
+import {getPath} from '../route-tree'
+
+type Props = {
+  folderBadge: number,
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/app/main.native.js]: nope 🔥

File: shared/actions/profile/index.js [hunk 12]
```diff
@@ -115,33 +112,6 @@ function _openURLIfNotNull(nullableThing, url, metaText): void {
   openURL(url)
 }
 
-function _onAppLink(action: ConfigGen.LinkPayload, state: TypedState) {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/profile/index.js]: ah this works on startup but i broke it while changing stuff around if you click while the app is alive

File: shared/app/routes-login.js (hunk 6 above)
Comment (chrisnojima) [reply shared/app/routes-login.js]: i added more comments but routetree typing is pretty broken

File: shared/engine/index.js (hunk 7 above)
Comment (chrisnojima) [reply shared/engine/index.js]: no, https://github.com/keybase/client/blob/master/shared/engine/index.js#L432

File: shared/engine/index.js (hunk 8 above)
Comment (chrisnojima) [reply shared/engine/index.js]: those are kind used interchangably, we never talk directly to the api server so in our case server = service

**Pull request merged**